src/mds/Server.cc

```diff
@@ -7928,6 +7928,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
   if (!dn)
     return;
 
+  CDentry::linkage_t *dnl = dn->get_linkage(client, mdr);
+  ceph_assert(!dnl->is_null());
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  auto* primary_diri = in->get_projected_parent_dir()->inode;
+  if (primary_diri->is_stray()) {
```

Comment:
Okay I've done some extensive testing today and have been unable to ever get this code path to be taken. The main issue is that the client will always `lookup` the file path before unlink. When that unlink request is cleaned up and unlocked, the Locker will `eval_remote` to initiate any reintegration. That will likely occur before the followup unlink request can be issued. Consequently, the rename acquires the locks that the `unlink` RPC needs so it's blocked until that rename completes. By that time, the `primary_diri` is no longer a stray.

I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?

Also, I've been looking at #47399 which introduced `WAIT_UNLINK`. After having looked at all of these PRs in depth, that original issue

https://tracker.ceph.com/issues/56695

indicates to me that the problem is 

https://github.com/ceph/ceph/blob/552b6d852cbc7384a05805989a675de7d6b4a3fd/src/mds/Server.cc#L7385

in that this is the **wrong** parent to look at. The parent dn of the inode is irrelevant. What matters is that the target dn's parent is in the same subvolume. This is precisely because the actual inode may temporarily be located in the stray directory pending reintegration. So I *think* that the right fix for issue 56695 is to simply check the snap realm of `srcdn` (which would be gotten from `rdlock_two_paths_xlock_destdn`). Alternatively, don't return EXDEV if the inode's parent dir is a stray.

What do you think? I'm leery of WAIT_UNLINK for the same reasons as WAIT_REINTEGRATE.

Reply:
> Okay I've done some extensive testing today and have been unable to ever get this code path to be taken. The main issue is that the client will always `lookup` the file path before unlink. When that unlink request is cleaned up and unlocked, the Locker will `eval_remote` to initiate any reintegration. That will likely occur before the followup unlink request can be issued. Consequently, the rename acquires the locks that the `unlink` RPC needs so it's blocked until that rename completes. By that time, the `primary_diri` is no longer a stray.
> 

From my reading the previous test failure logs, it seems the first `unlink` hasn't triggered the `linkmerge` when MDS received a second `unlink`. The test cases it will be something like:

Primary dentry path : /s1/d1/c1/file
Remote dentry path:  /s2/d2/c2/file_link1
Remote dentry path:  /s3/d3/c3/file_link2

Client issued the `unlink` for the primary denty, and then a second `unlink` for one of the remote dentry.

I didn't see why the lookup need to wait the first `unlink` request must release the non-read locks. Just after the early reply the `unlink` request will release the `rdlocks`, won't that enough ?

A detail example please see https://tracker.ceph.com/issues/61818#note-2.

> I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?
> 

Before I reproduce this with the `fsstress` test script. And also could see the test failure in `pjd` test script. I think you can just run the `fsstress` script to have a try.

> Also, I've been looking at #47399 which introduced `WAIT_UNLINK`. After having looked at all of these PRs in depth, that original issue
> 
> https://tracker.ceph.com/issues/56695
> 
> indicates to me that the problem is
> 
> https://github.com/ceph/ceph/blob/552b6d852cbc7384a05805989a675de7d6b4a3fd/src/mds/Server.cc#L7385
> 
> in that this is the **wrong** parent to look at. The parent dn of the inode is irrelevant. What matters is that the target dn's parent is in the same subvolume. This is precisely because the actual inode may temporarily be located in the stray directory pending reintegration. So I _think_ that the right fix for issue 56695 is to simply check the snap realm of `srcdn` (which would be gotten from `rdlock_two_paths_xlock_destdn`). Alternatively, don't return EXDEV if the inode's parent dir is a stray.
> 
> What do you think? I'm leery of WAIT_UNLINK for the same reasons as WAIT_REINTEGRATE.

Checked the original issue and the mds code again, yeah, this approach seems won't works because even the `linkmerge` is not triggered yet the `link` still could success and won't depend on it.

Let me update this PR to revert all the previous patches and run the test again.

Thanks @batrick